File: ConsultaPoliza.Api/Controllers/BranchesController.cs
using ConsultaPoliza.Api.Models;
using ConsultaPoliza.Api.Services;
using Microsoft.AspNetCore.Mvc;
using Oracle.ManagedDataAccess.Client;

namespace ConsultaPoliza.Api.Controllers;

[ApiController]
[Route("api/ramos")]
public sealed class BranchesController : ControllerBase
{
    private readonly IBranchRepository _branchRepository;
    private readonly ILogger<BranchesController> _logger;

    public BranchesController(IBranchRepository branchRepository, ILogger<BranchesController> logger)
    {
        _branchRepository = branchRepository;
        _logger = logger;
    }

    [HttpGet]
    [ProducesResponseType(typeof(IReadOnlyList<BranchResponse>), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status503ServiceUnavailable)]
    public async Task<ActionResult<IReadOnlyList<BranchResponse>>> GetAll(CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _branchRepository.GetAllAsync(cancellationToken));
        }
        catch (OracleException ex) when (IsOracleConnectionError(ex))
        {
            _logger.LogError(ex, "Oracle connection failed while loading branches.");
            return Problem(
                title: "Oracle no esta disponible.",
                detail: "No se pudo conectar con Oracle para cargar los ramos. Verifique la red o VPN.",
                statusCode: StatusCodes.Status503ServiceUnavailable);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error loading branches.");
            return Problem(
                title: "No se pudieron cargar los ramos.",
                detail: "Ocurrio un error tecnico al consultar los ramos en Oracle.",
                statusCode: StatusCodes.Status500InternalServerError);
        }
    }

    private static bool IsOracleConnectionError(OracleException exception)
    {
        return exception.Number is 50201 or 50232
            || exception.Errors.Cast<OracleError>().Any(error => error.Number is 50201 or 50232);
    }
}

File: ConsultaPoliza.Api/Controllers/BranchesController.vb
Imports System.Diagnostics
Imports System.Net
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Web.Http
Imports ConsultaPoliza.Api.Infrastructure
Imports ConsultaPoliza.Api.Models
Imports ConsultaPoliza.Api.Services
Imports Oracle.ManagedDataAccess.Client

Namespace ConsultaPoliza.Api.Controllers
    <RoutePrefix("api/ramos")>
    Public Class BranchesController
        Inherits ApiController

        Private ReadOnly _branchRepository As IBranchRepository

        Public Sub New()
            Me.New(ServiceRegistry.BranchRepository)
        End Sub

        Friend Sub New(branchRepository As IBranchRepository)
            _branchRepository = branchRepository
        End Sub

        <HttpGet>
        <Route("")>
        Public Async Function GetAll() As Task(Of IHttpActionResult)
            Try
                Dim branches = Await _branchRepository.GetAllAsync(CancellationToken.None)
                Return Ok(branches)
            Catch ex As OracleException When IsOracleConnectionError(ex)
                Trace.TraceError("Oracle connection failed while loading branches: {0}", ex)
                Return Content(
                    HttpStatusCode.ServiceUnavailable,
                    New ErrorResponse With {
                        .Title = "Oracle no esta disponible.",
                        .Detail = "No se pudo conectar con Oracle para cargar los ramos. Verifique la red o VPN."
                    })
            Catch ex As Exception
                Trace.TraceError("Error loading branches: {0}", ex)
                Return Content(
                    HttpStatusCode.InternalServerError,
                    New ErrorResponse With {
                        .Title = "No se pudieron cargar los ramos.",
                        .Detail = "Ocurrio un error tecnico al consultar los ramos en Oracle."
                    })
            End Try
        End Function

        Private Shared Function IsOracleConnectionError(exception As OracleException) As Boolean
            If exception.Number = 50201 OrElse exception.Number = 50232 Then
                Return True
            End If

            For Each [error] As OracleError In exception.Errors
                If [error].Number = 50201 OrElse [error].Number = 50232 Then
                    Return True
                End If
            Next

            Return False
        End Function
    End Class
End Namespace
